File: gcjPCPPA.hs
#!/usr/bin/env runghc
import Data.List
import Data.Maybe
main = getContents >>= mapM_ putStrLn . zipWith caseno [1..] . solve . lines
caseno i r = "Case #"++show i++": "++r
solve (c:cs) = map (toString . answer) . take (read c) . analyze $ cs ::[String]
analyze = map words
toString = id
answer [a,s,t] = reverse $ f $ foldl (\b c->b*n+fromJust(c`elemIndex`s)) 0 a
  where
    [n,m] = map length [s,t]
    f b = if b==0 then "" else t!!r:f d where (d,r) = b`divMod` m

File: gcjPCPPA.py
#!/usr/bin/env python
import sys

def solve():
    A,S,T = sys.stdin.readline().split()
    N,M = len(S),len(T)
    a = reduce(lambda a,c: a*N+S.index(c),A,0);
    t = ""
    while a!=0:
        t=T[a%M]+t
        a=a//M
    return t

for x in range(int(sys.stdin.readline())):
    print("Case #"+str(x+1)+": "+str(solve()))
